

<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  <title>Tombo Examples &mdash; Tombo 1.1.1 documentation</title>
  

  
  
  
  

  

  
  
    

  

  
  
    <link rel="stylesheet" href="_static/css/theme.css" type="text/css" />
  

  

  
        <link rel="index" title="Index"
              href="genindex.html"/>
        <link rel="search" title="Search" href="search.html"/>
    <link rel="top" title="Tombo 1.1.1 documentation" href="index.html"/>
        <link rel="next" title="Re-squiggle Algorithm" href="resquiggle.html"/>
        <link rel="prev" title="Welcome to Tombo’s documentation!" href="index.html"/> 

  
  <script src="_static/js/modernizr.min.js"></script>

</head>

<body class="wy-body-for-nav" role="document">

   
  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search">
          

          
            <a href="index.html" class="icon icon-home"> Tombo
          

          
          </a>

          
            
            
              <div class="version">
                1.1.1
              </div>
            
          

          
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>

          
        </div>

        <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
          
            
            
              
            
            
              <ul class="current">
<li class="toctree-l1 current"><a class="current reference internal" href="#">Tombo Examples</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#re-squiggle-raw-signal-alignment">Re-squiggle (Raw Signal Alignment)</a></li>
<li class="toctree-l2"><a class="reference internal" href="#modified-base-detection">Modified Base Detection</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#specific-alternative-base-detection">Specific Alternative Base Detection</a></li>
<li class="toctree-l3"><a class="reference internal" href="#canonical-sample-comparison-detection">Canonical Sample Comparison Detection</a></li>
<li class="toctree-l3"><a class="reference internal" href="#de-novo-non-canonical-base-detection">De novo Non-canonical Base Detection</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#text-output">Text Output</a><ul>
<li class="toctree-l3"><a class="reference internal" href="#wiggle-format-output">Wiggle Format Output</a></li>
<li class="toctree-l3"><a class="reference internal" href="#genome-sequence-output">Genome Sequence Output</a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#plotting-examples">Plotting Examples</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="resquiggle.html">Re-squiggle Algorithm</a></li>
<li class="toctree-l1"><a class="reference internal" href="modified_base_detection.html">Modified Base Detection</a></li>
<li class="toctree-l1"><a class="reference internal" href="text_output.html">Text Outputs</a></li>
<li class="toctree-l1"><a class="reference internal" href="plotting.html">Plotting Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="filtering.html">Read Filtering Commands</a></li>
<li class="toctree-l1"><a class="reference internal" href="model_training.html">Model Training (Advanced Users Only)</a></li>
</ul>
<ul>
<li class="toctree-l1"><a class="reference internal" href="tombo.html">tombo package</a></li>
</ul>

            
          
        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">Tombo</a>
        
      </nav>


      
      <div class="wy-nav-content">
        <div class="rst-content">
          















<div role="navigation" aria-label="breadcrumbs navigation">

  <ul class="wy-breadcrumbs">
    
      <li><a href="index.html">Docs</a> &raquo;</li>
        
      <li>Tombo Examples</li>
    
    
      <li class="wy-breadcrumbs-aside">
        
            
            <a href="_sources/examples.rst.txt" rel="nofollow"> View page source</a>
          
        
      </li>
    
  </ul>

  
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
            
  <div class="section" id="tombo-examples">
<h1>Tombo Examples<a class="headerlink" href="#tombo-examples" title="Permalink to this headline">¶</a></h1>
<p>Below are minimal use case examples. For more detail on each commands options and further algorithm details, please see the corresponding documentation sections.</p>
<div class="section" id="re-squiggle-raw-signal-alignment">
<h2>Re-squiggle (Raw Signal Alignment)<a class="headerlink" href="#re-squiggle-raw-signal-alignment" title="Permalink to this headline">¶</a></h2>
<p>The re-squiggle algorithm aligns raw signal to genomic sequence based on a genomic mapping.</p>
<p>This command will add infomation including the mapped genomic location and the raw signal to sequence assignment into the FAST5 files provided as well as producing an index file for more efficient file access in downstream commands.</p>
<p>The <code class="docutils literal"><span class="pre">resquiggle</span></code> command must be run on any set of FAST5s before any further processing by Tombo.</p>
<p><strong>Important note</strong>: Currently, only a models for R9.4/5 sequencing DNA or RNA are included with Tombo. Analysis of other nanopore data types is not supported at this time.</p>
<p>For more details see the <a class="reference internal" href="resquiggle.html"><span class="doc">re-squiggle documentation</span></a>.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span><span class="c1"># optionally annotate raw FAST5s with FASTQ files produced from the same reads</span>
tombo annotate_raw_with_fastqs --fast5-basedir &lt;fast5s-base-directory&gt; --fastq-filenames reads.fastq

tombo resquiggle <span class="o">[</span>-h<span class="o">]</span> &lt;fast5s-base-directory&gt; &lt;reference-fasta&gt; --minimap2-executable ./minimap2
</pre></div>
</div>
</div>
<div class="section" id="modified-base-detection">
<h2>Modified Base Detection<a class="headerlink" href="#modified-base-detection" title="Permalink to this headline">¶</a></h2>
<p>Tombo provides three methods for the investigation of modified bases. Each method has different advantages and requirements.</p>
<ul class="simple">
<li>The specific alternative base method is preferred, but is currently only available for 5mC in DNA (more modifications coming soon).</li>
<li>The canonical (control) sample comparison method would be preferred next, but requires the producetion of a second set of reads containing only the 4 canonical bases.</li>
<li>The de novo method is recommended only a research tool and may produce high false positive rates.</li>
<li>Additionally, both the control sample and de novo methods may not identify the exact modified base location and give no information as to the identity of a modified base.</li>
</ul>
<p>The result of all <code class="docutils literal"><span class="pre">test_significance</span></code> calls will be a binary statistics file, which can be passed to other Tombo sub-commands.</p>
<p>For more details see the <a class="reference internal" href="modified_base_detection.html"><span class="doc">modified base detection documentation</span></a>.</p>
<div class="section" id="specific-alternative-base-detection">
<h3>Specific Alternative Base Detection<a class="headerlink" href="#specific-alternative-base-detection" title="Permalink to this headline">¶</a></h3>
<p>In order to specifically detect 5-methyl cytosine (and other alternative bases in the future), use the <code class="docutils literal"><span class="pre">test_significance</span></code> command with the <code class="docutils literal"><span class="pre">--alternate-bases</span> <span class="pre">5mC</span></code> option.</p>
<p>This will perform a log likelihood ratio test using the default canonical and 5mC alternative models provided with Tombo.</p>
<p>New alternative base models will be added as they are trained. This is the perferred method for modified base detection if a model is available for your biological sample of interest.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>tombo test_significance --fast5-basedirs &lt;fast5s-base-directory&gt; <span class="se">\</span>
    --alternate-bases 5mC --statistics-file-basename sample_5mC_detection
</pre></div>
</div>
</div>
<div class="section" id="canonical-sample-comparison-detection">
<h3>Canonical Sample Comparison Detection<a class="headerlink" href="#canonical-sample-comparison-detection" title="Permalink to this headline">¶</a></h3>
<p>In order to perform canonical-sample-comparison modified base detection, use the <code class="docutils literal"><span class="pre">test_significance</span></code> command with a second set of reads from the same biological sample containing only canonical bases (e.g. PCR) using the <code class="docutils literal"><span class="pre">--control-fast5-basedirs</span></code> option.</p>
<p>This will perform a hypothesis test against the signal level observed from the control sample at each genomic position. This method provides the highest accuracy, but does not always identify the exact modification position or the identity of the modified base.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>tombo test_significance --fast5-basedirs &lt;fast5s-base-directory&gt; <span class="se">\</span>
    --control-fast5-basedirs &lt;control-fast5s-base-directory&gt;  <span class="se">\</span>
    --statistics-file-basename sample_canonical_compare
</pre></div>
</div>
</div>
<div class="section" id="de-novo-non-canonical-base-detection">
<h3>De novo Non-canonical Base Detection<a class="headerlink" href="#de-novo-non-canonical-base-detection" title="Permalink to this headline">¶</a></h3>
<p>In order to perform de novo non-canonical base detection, use the <code class="docutils literal"><span class="pre">test_significance</span></code> command without any other options (aside from the set of reads to test).</p>
<p>This will perform a hypothesis test against the default canonical base model provided with Tombo. Note that this method is quite error prone and will likely result in a high number of false positives, but may be of use in a research and development setting. This method also has the lowest requirement of only a set of reads and a genome.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>tombo test_significance --fast5-basedirs &lt;fast5s-base-directory&gt; <span class="se">\</span>
    --statistics-file-basename sample_de_novo_detection
</pre></div>
</div>
</div>
</div>
<div class="section" id="text-output">
<h2>Text Output<a class="headerlink" href="#text-output" title="Permalink to this headline">¶</a></h2>
<div class="section" id="wiggle-format-output">
<h3>Wiggle Format Output<a class="headerlink" href="#wiggle-format-output" title="Permalink to this headline">¶</a></h3>
<p>In order to output the results of re-squiggling and statistical testing in a genome browser compatible format (<a class="reference external" href="https://genome.ucsc.edu/goldenpath/help/wiggle.html">wiggle format</a>), the <code class="docutils literal"><span class="pre">write_wiggles</span></code> subcommand is provided.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>tombo write_wiggles --fast5-basedirs &lt;fast5s-base-directory&gt; --wiggle-basename sample_5mC_detection <span class="se">\</span>
    --statistics-filename sample_5mC_detection.5mC.tombo.stats --wiggle-types fraction coverage
</pre></div>
</div>
</div>
<div class="section" id="genome-sequence-output">
<h3>Genome Sequence Output<a class="headerlink" href="#genome-sequence-output" title="Permalink to this headline">¶</a></h3>
<p>For modified base analysis pipelines (e.g. motif detection), it may be of use to output the genomic sequence surrounding locations with the largest fraction of modified reads. The <code class="docutils literal"><span class="pre">write_most_significant_fasta</span></code> sub-command is provided for this purpose.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>tombo write_most_significant_fasta --statistics-filename sample_5mC_detection.5mC.tombo.stats <span class="se">\</span>
    --genome-fasta &lt;reference-fasta&gt;
</pre></div>
</div>
<p>Example <a class="reference external" href="http://meme-suite.org/doc/meme.html">meme</a> command line modified base motif detection command.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>./meme -oc motif_output.meme -dna -mod zoops tombo_results.significant_regions.fasta
</pre></div>
</div>
<p>For more details see the <a class="reference internal" href="text_output.html"><span class="doc">text output documentation</span></a>.</p>
</div>
</div>
<div class="section" id="plotting-examples">
<h2>Plotting Examples<a class="headerlink" href="#plotting-examples" title="Permalink to this headline">¶</a></h2>
<p>Tombo provides many plotting functions for the visualization of potentially modified bases and the raw nanopore signal in general.</p>
<p>Most plotting commands are genome-anchored. That is the raw signal is plotted as the re-squiggle algorithm has assigned it to the genome. Thus each read contain a different number of raw observations per genomic base. For summary distributions (not raw signal) the distributions are taken over each reads average signal level at the genomic position.</p>
<p>Each genome anchored plotting command allows for the selection of genomic positions based on generally applicable criterion.</p>
<div class="highlight-bash"><div class="highlight"><pre><span></span>tombo plot_max_coverage --fast5-basedirs &lt;fast5s-base-directory&gt; --plot-standard-model

tombo plot_motif_centered --fast5-basedirs &lt;fast5s-base-directory&gt; --motif AWC <span class="se">\</span>
    --genome-fasta genome.fasta --control-fast5-basedirs &lt;control-fast5s-base-directory&gt;

tombo plot_per_read --fast5-basedirs &lt;fast5s-base-directory&gt; <span class="se">\</span>
    --genome-locations chromosome:1000 chromosome:2000:- --plot-alternate-model 5mC
</pre></div>
</div>
<p>For more details see the <a class="reference internal" href="plotting.html"><span class="doc">plotting documentation</span></a>.</p>
<blockquote>
<div>For additional command details, see the specific commands documentation section.</div></blockquote>
</div>
</div>


           </div>
           <div class="articleComments">
            
           </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
        <a href="resquiggle.html" class="btn btn-neutral float-right" title="Re-squiggle Algorithm" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right"></span></a>
      
      
        <a href="index.html" class="btn btn-neutral" title="Welcome to Tombo’s documentation!" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <p>
        &copy; Copyright 2017, Oxford Nanopore Technologies.

    </p>
  </div>
  Built with <a href="http://sphinx-doc.org/">Sphinx</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>. 

</footer>

        </div>
      </div>

    </section>

  </div>
  


  

    <script type="text/javascript">
        var DOCUMENTATION_OPTIONS = {
            URL_ROOT:'./',
            VERSION:'1.1.1',
            COLLAPSE_INDEX:false,
            FILE_SUFFIX:'.html',
            HAS_SOURCE:  true,
            SOURCELINK_SUFFIX: '.txt'
        };
    </script>
      <script type="text/javascript" src="_static/jquery.js"></script>
      <script type="text/javascript" src="_static/underscore.js"></script>
      <script type="text/javascript" src="_static/doctools.js"></script>
      <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

  

  
  
    <script type="text/javascript" src="_static/js/theme.js"></script>
  

  
  
  <script type="text/javascript">
      jQuery(function () {
          SphinxRtdTheme.StickyNav.enable();
      });
  </script>
   

</body>
</html>